### ISSUE TITLE
Using `nonce` to improve handling of CSP / CSP improvements

### ISSUE THREAD

**Issue by Witoso** [ckeditor/ckeditor5#16791]
1. ✅  Assigning styles to elements in engine. It was mentioned by @oleq in this [comment](https://github.com/ckeditor/ckeditor5/issues/15509#issuecomment-1876800147). I discovered that assigning styles to `.cssText` bypasses this issue without much modification of the code, so I think it's fixed. [Here is PoC](https://github.com/ckeditor/ckeditor5/pull/16728).  
2. ⚠️ Pasting content with styles throws CSP exception. It happens in the HTML data processor right [there](https://github.com/ckeditor/ckeditor5/blob/ck/15509/packages/ckeditor5-engine/src/dataprocessor/htmldataprocessor.ts#L122) because we are attending to parse `<p style="color: red">Foo</p>` from clipboard. It's not fixed in PoC, but with a lot of hacks it might be. 
3. ❌  Let's assume we have `#editor` div with `<p style="color: red">Foo</p>` content. Calling `ClassicEditor.create( document.querySelector( '#editor' ), {` creates editor with `Foo` paragraph that is unstyled on Firefox. It's black instead of red. On the Chrome on the other hand it's working, `style` tag is preserved. It looks like Firefox disallows the browser to read `style` tag and trims these tags from HTML content before attending to bootstrap editor. I'm not sure if we can fix this.

I think that there are at least two ways to fix that issue:

1. ❓ Using [nonce](https://content-security-policy.com/nonce/) attribute on our elements affected by `style-src 'self'` CSP attribute. We can add an optional editor configuration option to specify the value of this attribute tag.  This approach is used by[ `styled-loader` ](https://webpack.js.org/loaders/style-loader/#nonce)used by `Webpack`. 
2. ❓ Keep my [PoC](https://github.com/ckeditor/ckeditor5/pull/16728) and try to bypass the issue in HTML parser. I'm not quite sure how to do that without writing a small HTML parser that walks and removes all `style` attributes, which applies styling in later phase using `cssText`.  It will not fix `3` issue. 

@Reinmar What do you think about my `1` proposition?

_Originally posted by @Mati365 in https://github.com/ckeditor/ckeditor5/issues/15509#issuecomment-2230126976_

**Comment by aljaznuncic** [comment]
Is using CSP nonce currently possible with custom build CKEditor5 (version 41 or later) and Angular 18? Any useful link for this?

**Comment by jeandat** [comment]
> Is using CSP nonce currently possible with custom build CKEditor5 (version 41 or later) and Angular 18? Any useful link for this?
> 
> My CSP HEADER at nginx: `add_header Content-Security-Policy "default-src 'self'; connect-src 'self'; img-src 'self' data:; font-src 'self' data:; style-src 'self' 'nonce-$request_id'; script-src 'self' 'nonce-$request_id'; frame-ancestors 'none'; object-src 'none'; frame-src 'self'" always; `
> 
> I got this: ![image](https://private-user-images.githubusercontent.com/39697469/355157724-cf60bb8b-d332-4413-9651-0bbe518e735b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYwNDI3MDQsIm5iZiI6MTcyNjA0MjQwNCwicGF0aCI6Ii8zOTY5NzQ2OS8zNTUxNTc3MjQtY2Y2MGJiOGItZDMzMi00NDEzLTk2NTEtMGJiZTUxOGU3MzViLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTExVDA4MTMyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWEyMTQ2MGJkYzAyZWVmNzg5NzA2MmYwNTJlZWY3ZjZkYTEyMDJlYjhiZWZiMWY3OTNhMjAyYzRiZTU5NjQ1MWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.5LhH3YKhLJIr5d-j84FMqF9GtlVOMZM7PTI83__ENro)

I don't think so mate since this ticket is still open. I also don't saw anything in documentation about configuring a nonce.

For the record I'm also very interested by this feature. Because of libraries like ckeditor, I cannot remove "unsafe-inline" and use a nonce on styles.

**Comment by jcompagner** [comment]
Yes i think CKEditor should just fix it as Angular also fixed this, as an example:

https://www.innoq.com/en/blog/2023/07/style-nonces-for-csp-in-angular-16/

so give use a config setting that we provide to the ckeditor instance, that us then used by ckeditor to set on the inline style tags that they generate

**Comment by Reinmar** [comment]
I'm a bit confused by this ticket and I think that it misses a very important thing.

CKEditor **does not generate inline `<style>` elements anymore**. 

It did so in the so called legacy installation methods. Those were the only installation methods available prior to v42.0 (June 2024). We used webpack style loader which was automatically inject `<style>` elements to the page.

However, since v42.0, it's the integrator's job to include CKEditor's stylesheets on the page. So, the integrator can use `nonce`. Check out the [latest docs](https://ckeditor.com/docs/ckeditor5/latest/getting-started/index.html).

Note: The legacy installation methods are still supported too. So, if you use v42.0+ you might still be using them. You might need to [migrate to the new installation methods](https://ckeditor.com/docs/ckeditor5/latest/updating/nim-migration/migration-to-new-installation-methods.html).

Do I miss something?

**Comment by Witoso** [comment]
Yep, I will update the title and description. Somehow I thought `nonce` could be used in the elements other than `style` and `script` but this is not the case.

**Comment by Witoso** [comment]
Thanks for pointing this out, I will update the docs.

**Comment by kglickman** [comment]
Unfortunately, I am not allowed to use  'unsafe-inline' on style-src-attr any more than I am on style-src so I've had to disable many plugins.

What would be helpful would be to make variations of the plugins that are slightly less flexible so that they could be styled with classes instead.  For instance, instead of allowing the selection of an infinite number of colors, allow for a subset so that you could apply a class named 'red' instead of style with color:#FF0000.  For fonts, have a default list of fonts so that you could have a class named 'arial'.  For things like image sizing, instead of supporting widths with pixels, support classes for various percentages.  For indents, allow styles for up to 10 indents.  You get the idea.

**Comment by iamnexxed** [comment]
Removing the use of `unsafe-inline` would be a great advantage as I need to use strictest CSP. @kglickman has a decent suggestion for the solution.

**Comment by apoorv-2204** [comment]
I am using nonce in my project to whitelist the css style tag.
is there any way that can cke editor5 generate css inside <style> tag inside of inline styles by defining a class for the element.
and for the style tag we can set nonce <style nonce={get_nonce()}.

Or is there any plugin that generate classes instead of inline styles?

**Comment by herrardo** [comment]
Is there any documentation that shows what tooling is generating inline styles?

As far as i have tested, some of the plugins change classes for the elements to alter its styling, which is perfectly fine for CSP, but others, as you are pointing out, are adding inline styles, so my question here would be if there are a documentation where it is shown which plugins add inline styling to prevent customers to use them in order to use CSP without unsafe- values.

On top of that documentation, some approach as @kglickman suggested would work to safley uise the plugins.

**Comment by jondeodat** [comment]
Hi guys, any news on that front? Security team is poking me every year on this topic. It appears in audits and some customers are also complaining. I think ckeditor is the last remaining bit in the whole app that still rely on 'unsafe-inline'. Have you made some progress? Is it a priority for you? Do you plan on doing something or is it in your backlog somewhere with no clear priority…

**Comment by Witoso** [comment]
No specific plans on your side but thanks for the ping. What would be useful, though, is to get an understanding of features that are in use. Such a topic would have a big cost if we consider all features with `style` output.

**Comment by kglickman** [comment]
We are using the list below and disable ['fontBackgroundColor', 'fontColor', 'fontFamily'] if the CSP does not have the 'unsafe-inline'.  We don't disable images or tables but some of the functionality for them doesn't work.
```
import {
	ClassicEditor, Alignment, Bold, Italic, Underline, Strikethrough, Subscript, Superscript,
	Code, BlockQuote, Essentials, FindAndReplace, FontBackgroundColor,
	FontColor, FontFamily, FontSize, Heading, Highlight, HorizontalLine,
	Image,
	ImageBlock,
	ImageCaption,
	ImageInline,
	ImageInsert,
	ImageResize,
	ImageStyle,
	ImageTextAlternative,
	ImageToolbar,
	ImageUpload,
	Indent, IndentBlock,
	Link, List, TodoList, Paragraph, RemoveFormat, SimpleUploadAdapter, SpecialCharacters, SpecialCharactersArrows, SpecialCharactersCurrency, SpecialCharactersEssentials,
	SpecialCharactersLatin, SpecialCharactersMathematical, SpecialCharactersText, SourceEditing,
	Table, TableToolbar, TableCaption, TableCellProperties, TableColumnResize, TableProperties, PasteFromOffice, Fullscreen
} from 'ckeditor5';
```
